caffe2/core/operator_c10wrapper.h

```diff
@@ -25,104 +16,100 @@ using extract_type_t = typename ParameterDef::type;
  public:
-  static_assert(
-      c10::guts::is_instantiation_of<std::tuple, ParameterDefTuple>::value,
-      "");
-  using ParameterTuple =
-      c10::guts::typelist::to_tuple_t<c10::guts::typelist::map_t<
-          details::extract_type_t,
-          c10::guts::typelist::from_tuple_t<ParameterDefTuple>>>;
-
   USE_OPERATOR_CONTEXT_FUNCTIONS;
 
-  C10OperatorWrapper(const OperatorDef& operator_def, Workspace* ws)
+  C10OperatorWrapper(
+      const c10::OperatorHandle& op,
+      const OperatorDef& operator_def,
+      Workspace* ws)
       : Operator<Context>(operator_def, ws),
-        op_(OperatorHandle()),
+        op_(op),
         kernel_(at::nullopt),
-        parameters_(parse_parameters_(
-            operator_def,
-            c10::guts::make_index_sequence<num_parameters()>())) {
+        has_preallocated_outputs_(
+            op_.schema().arguments().back().name() ==
+            detail::PREALLOCATED_OUTPUT_ARGNAME) {
+    AT_ASSERT(
+        !has_preallocated_outputs_ ||
+        op_.schema().arguments().back().type()->isSubtypeOf(
+            OptionalType::create(ListType::ofTensors())));
 
     AT_ASSERT(operator_def.output_size() == op_.schema().returns().size());
-    AT_ASSERT(operator_def.input_size() == num_inputs());
-  }
-
-  size_t num_inputs() {
-    return op_.schema().arguments().size() - 1 - num_parameters();
-  }
-
-  static constexpr size_t num_parameters() {
-    return std::tuple_size<ParameterDefTuple>::value;
+    AT_ASSERT(
+        operator_def.input_size() + (has_preallocated_outputs_ ? 1 : 0) <=
+        op_.schema()
+            .arguments()
+            .size()); // '<=' because there might be caffe2 arguments
   }
 
   bool RunOnDevice() override {
     // due to caching the stack_, concurrent calling is not allowed.
     // TODO thread_local might fix this
     std::lock_guard<std::mutex> lock(mutex_);
 
-    AT_ASSERT(stack_.size() == 0);
-
     pushInputs_();
-    pushParameters_(guts::make_index_sequence<num_parameters()>());
-    pushOutputParameters_();
-
     callKernel_();
-
     popOutputs_();
 
     return true;
   }
 
  private:
-  template <size_t... ParameterIndex>
-  ParameterTuple parse_parameters_(
-      const OperatorDef& operator_def,
-      c10::guts::index_sequence<ParameterIndex...>) {
-    return ParameterTuple{Parameter<ParameterIndex>(operator_def)...};
-  }
-
-  template <size_t Index>
-  details::extract_type_t<
-      typename std::tuple_element<Index, ParameterDefTuple>::type>
-  Parameter(const OperatorDef& operator_def) {
-    using Parameter =
-        typename std::tuple_element<Index, ParameterDefTuple>::type;
-    return Parameter::parse(ArgumentHelper(operator_def));
-  }
-
   void pushInputs_() {
-    if (use_array_input) {
-      stack_.emplace_back(ivalue::TensorList::create(array_inputs_()));
-    } else {
-      for (size_t i = 0; i < num_inputs(); ++i) {
-        stack_.emplace_back(at::Tensor(C10Tensor(Input(i))));
+    AT_ASSERT(stack_.size() == 0);
```

Comment:
Umm, why? The whole point of a stack machine is that you don't have to allocate a fresh argument vector every time you make a call; you can use the suffix to hold whatever set of arguments you need for the op you're running (but the prefix gets ignored.)